core: introduce support for cgroup freezer

msekletar opened the pull request:
With cgroupv2 the cgroup freezer is implemented as a cgroup
attribute (knob) called cgroup.freeze. cgroup can be frozen by writing
"1" to the file and kernel will send us a notification through
"cgroup.events" after the operation is finished and processes in the
cgroup entered quiescent state, i.e. they are not scheduled to
run. Writing "0" to the attribute file does the inverse and process
execution is resumed. See [1] for more details of the new interface.

This commit exposes above low-level functionality through systemd's DBus
API. Two new methods on each unit object are introduced, Freeze() and
Thaw(). However, each unit type must provide specialized implementation
for these methods, otherwise, we return an error (EOPNOTSUPP) to the
client requesting the operation on the unit type w/o the support. So far
only service, scope, and slice unit have support.
Note that DBus API has a synchronous behavior and we dispatch the reply
to freeze/thaw requests only after the kernel has notified us that
requested operation was completed. systemctl verbs that wrap the dbus
calls are also provided.

Implementation is accompanied by very basic integration test.

[1] https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html?highlight=freezer#core-interface-files

msekletar commented:
CI fails because I've messed up test-suite integration. I was running the test case manually and I should have tried to run it as part of the overall test-suite. I will update the PR with the fix.

mrc0mmand commented:
The fail on Arch is most likely a timing issue (as it failed with `Call failed: Connection timed out`). I guess tweaking the `busctl` timeout via `--timeout` might help in this case, as the test runs in a VM without KVM acceleration.

The issue would be, probably, present in Ubuntu CI as well, but in this case the test is skipped there:

```
+ set -e
+ set -o pipefail
+ unit=foo.service
+ unit_file=/etc/systemd/system/foo.service
+ test -e /sys/fs/cgroup/system.slice/cgroup.freeze
+ echo OK
+ exit 0
```

mrc0mmand commented:
Also, would it make sense to *force* cgroups v2 for this particular test? Something along:

```
KERNEL_APPEND="$KERNEL_APPEND systemd.unified_cgroup_hierarchy=true"
```

in the `test.sh` file, as we, sort of, already do with the SELinux test:

https://github.com/systemd/systemd/blob/b1ccd384d6d3524484f0153ab23eda8ad67b476a/test/TEST-06-SELINUX/test.sh#L16

This way the test could test what it's supposed to even on systems which default to legacy/hybrid hierarchies.

mrc0mmand commented on test/TEST-38-FREEZER/testsuite.sh:
```diff
@@ -0,0 +1,194 @@
+
+set -e
+set -o pipefail
+
```
Also, to make the debugging much easier (especially now with the `busctl` timeout), it may be worthwhile to enable debug logging and log it directly to the console, like we do in other tests:

https://github.com/systemd/systemd/blob/b1ccd384d6d3524484f0153ab23eda8ad67b476a/test/TEST-23-TYPE-EXEC/testsuite.sh#L5-L6

msekletar replied:
Fixed

boucman commented:
would it be possible to have a "systemctl is-frozen" ?  is this visible on "systemctl status" ?

msekletar commented:
Yes, it is visible in systemctl output. 

```
[msekleta@f30 ~]$ sudo systemctl freeze cups
[sudo] password for msekleta:
[msekleta@f30 ~]$ sudo systemctl status cups
● cups.service - CUPS Scheduler
   Loaded: loaded (/usr/lib/systemd/system/cups.service; disabled; vendor preset: disabled)
   Active: active (frozen) since Fri 2019-09-20 18:05:29 CEST; 26s ago
     Docs: man:cupsd(8)
 Main PID: 1472 (cupsd)
   Status: "Scheduler is running..."
    Tasks: 1
   Memory: 6.2M
      CPU: 27ms
   CGroup: /system.slice/cups.service
           └─1472 /usr/sbin/cupsd -l

Sep 20 18:05:29 f30 systemd[1]: Starting CUPS Scheduler...
Sep 20 18:05:29 f30 systemd[1]: Started CUPS Scheduler.
```

cdown commented on man/systemctl.xml:
```diff
@@ -943,6 +943,29 @@ Sun 2017-02-26 20:57:49 EST  2h 3min left  Sun 2017-02-26 11:56:36 EST  6h ago
+            <para>Freeze one or more units specified on the
+            command line.</para>
+
+            <para>Freeze the units using cgroup freezer. Freezing the unit will cause that all processes contained within
```
The grammar here is not so great:

> Freezing the unit will cause all processes contained within the cgroup corresponding to the unit to be stopped until the unit is explicitly thawed.

msekletar replied:
Thank you for the suggestion.

msekletar replied:
This is fixed in the latest version.

cdown commented on src/core/cgroup.h:
```diff
@@ -32,6 +32,23 @@ typedef enum CGroupDevicePolicy {
+typedef enum CGroupControlEvent {
+        CGROUP_EMPTY,
+        CGROUP_FROZEN,
+        CGROUP_THAW,
```
Hmm, shouldn't this be `CGROUP_THAWED`, since it indicates an end state?

msekletar replied:
Agreed.

msekletar replied:
Resolved in latest version

cdown commented on src/core/dbus-unit.c:
```diff
@@ -722,6 +723,66 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
+
+        r = unit_freeze(u);
+        if (r == -EOPNOTSUPP)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Unit '%s' does not supporting freezing.", u->id);
```
This is misleading -- this implies that this unit will never support freezing, but it's not necessarily the case. It could also be that the cgroup is not set up properly yet (from unit_cgroup_freezer_action), that we're not on unified hierarchy (so it's not unit-specific), etc...

Either that or consider EOPNOTSUPP vs. EBUSY (I'll leave you to think about semantics here, but maybe it at least allows to distinguish the specificity of the issue)

msekletar replied:
Well, the message says that the unit doesn't support freezing at the time of the request. I don't see there any future implication. It is always true for every unit on cgroup v1. With v2 we support freezing only when unit type has freezing implemented. Also units cgroup must be already in place at the time we get Freeze/Thaw request since the `u->cgroup_path` is established before we even fork-off the service binary (https://github.com/systemd/systemd/blob/master/src/core/service.c#L1458)

However, I should add the case for EBUSY. I feel like the error handling for freezing should be overall similar to what we do in `bus_unit_method_clean()`.

msekletar replied:
Reference to `service.c` is not entirely relevant since the freezing is supported also by slices and scopes. But I think that you can't have a unit that is active (we support freezing only active units) and cgroup that is not realized.

cdown commented on src/core/dbus-unit.c:
```diff
@@ -722,6 +723,66 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
+
+        r = unit_thaw(u);
+        if (r == -EOPNOTSUPP)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Unit '%s' does not supporting freezing.", u->id);
```
ditto my above comment

msekletar commented:
I was playing with the failing test case and I've figured out that timeout happens because apparently dbus reply about the unit being frozen was not sent. Previously in `unit_frozen()` I've added the unit to dbus queue so once it is dispatched in next event loop iteration we send out the pending reply in `bus_unit_send_pending_freezer_message()`.

Now I've changed the code so that reply is sent directly from `unit_frozen()`. I don't think that it is the best solution but it makes test case to PASS. I will keep it for now, until I figure out some better fix.

mrc0mmand commented on src/core/dbus-unit.c:
```diff
@@ -722,6 +723,53 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
 }
 
+static int bus_unit_method_freezer_generic(sd_bus_message *message, void *userdata, sd_bus_error *error, FreezerAction action) {
+        const char* perm;
```
```
    archlinux_systemd_ci: ../src/core/dbus-unit.c: In function ‘bus_unit_method_freezer_generic’:
    archlinux_systemd_ci: ../src/core/dbus-unit.c:727:21: error: variable ‘perm’ set but not used [-Werror=unused-but-set-variable]
    archlinux_systemd_ci:   727 |         const char* perm;
    archlinux_systemd_ci:       |                     ^~~~
    archlinux_systemd_ci: cc1: all warnings being treated as errors
```

mrc0mmand replied:
This is caused by `mac_selinux_unit_access_check()` being a macro, which is optimized away if we compile without SELinux support (which we do):

https://github.com/systemd/systemd/blob/9d78dd1cce0f7f78153e38c6e867f31e0eb347b3/src/core/selinux-access.h#L11-L24

So, basically, the `r = mac_selinux_unit_access_check(u, message, perm, error);` line expands to `r = 0;`, which effectively makes the `perm` variable to be set, but not used.

msekletar replied:
I've pushed additional commit which removes the preprocessor check in the header as it is not really needed and as a consequence, the warning should go away.

msekletar commented:
CI si finally all green. @cdown I'd appreciate if you could have another look.

cdown commented on src/core/dbus-unit.c:
```diff
@@ -722,6 +723,53 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
+        if (r == -EOPNOTSUPP)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Unit '%s' does not supporting freezing.", u->id);
+        if (r == -EBUSY)
+                return sd_bus_error_setf(error, BUS_ERROR_UNIT_BUSY, "Unit is not inactive or has pending job.");
```
has pending job -> has a pending job

cdown commented on src/core/selinux-access.h:
```diff
@@ -8,17 +8,8 @@
 
 int mac_selinux_generic_access_check(sd_bus_message *message, const char *path, const char *permission, sd_bus_error *error);
 
-#if HAVE_SELINUX
```
Huh? Where did this go?

cdown commented on src/core/scope.c:
```diff
@@ -571,6 +571,30 @@ static const char* const scope_result_table[_SCOPE_RESULT_MAX] = {
         [SCOPE_FAILURE_TIMEOUT] = "timeout",
 };
 
+static int scope_freeze(Unit *s) {
```
These seem to be exactly the same for scope, service, whatever -- we should just create unit_freeze_common and unit_thaw_common

cdown commented on src/core/service.h:
```diff
@@ -116,7 +116,7 @@ struct Service {
         CGroupContext cgroup_context;
 
-        ServiceState state, deserialized_state;
+        ServiceState state, deserialized_state, frozen_state;
```
I don't see this used anywhere else, is it a leftover?

cdown commented on src/core/slice.c:
```diff
@@ -349,6 +350,55 @@ static void slice_enumerate_perpetual(Manager *m) {
+        _cleanup_closedir_ DIR *d = NULL;
+
+        assert(s);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_THAW));
```
ditto on crashing pid 1

cdown commented on src/core/unit.c:
```diff
@@ -721,6 +721,12 @@ void unit_free(Unit *u) {
         free(u);
 }
 
+FreezerState freezer_state(Unit *u) {
```
`unit_freezer_state`, judging by convention for other state retrievals that take a unit (like unit_sub_state)

cdown commented on src/core/unit.c:
```diff
@@ -3449,6 +3459,8 @@ int unit_serialize(Unit *u, FILE *f, FDSet *fds, bool serialize_jobs) {
         if (!sd_id128_is_null(u->invocation_id))
                 (void) serialize_item_format(f, "invocation-id", SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(u->invocation_id));
 
+        (void) serialize_item_format(f, "freezer-state", "%d", u->freezer_state);
```
This feels like it should be serialised as a string. I mean, if I see "freezer-state: 3", I have no idea what that means.

cdown commented on src/core/unit.c:
```diff
@@ -5962,6 +5979,73 @@ int unit_can_clean(Unit *u, ExecCleanMask *ret) {
+        int (*method)(Unit*);
+
+        assert(u);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_THAW));
```
ditto on conditional vs assert

cdown commented on src/systemctl/systemctl.c:
```diff
@@ -7717,6 +7777,8 @@ static int systemctl_help(void) {
                "  clean UNIT...                       Clean runtime, cache, state, logs or\n"
                "                                      or configuration of unit\n"
+               "  freeze UNIT...                      Freeze the unit using cgroup freezer\n"
+               "  thaw UNIT...                        Resume execution of the frozen unit\n"
```
Freeze *a* unit, resume execution of *a* frozen unit

Werkov commented on src/core/cgroup.c:
```diff
@@ -2801,8 +2853,16 @@ static int on_cgroup_inotify_event(sd_event_source *s, int fd, uint32_t revents,
-                        if (u)
-                                unit_add_to_cgroup_empty_queue(u);
+                        if (u) {
+                                event = cgroup_control_get_last_event(u);
```
Good catch on noticing the shared notifications. Furthermore, notifications can be merged not only in "space" but also in time (throttling).
I.e. we don't necessarily process the *last event* rather the last observed state.

My suggestion would be to implement sort of `cgroup_control_get_state` function that'd return flags for individual substates (i.e. populated, frozen). The actions triggered by the notifications shall be both idempotent and invertible, thus we'd always "apply the state" observed at the time of the most recent notification. (I.e. there is already `cgroup_{frozen|thawed}`, the function to deque unit from empty queue should be simple.)

Although I don't think PID 1 induced events can realistically mix up in such a pathological way, the "state observation" seems more robust should the contents of `cgroup.events` change in the future.

Werkov replied:
I've pushed [freezer-reading-events-expanded](https://github.com/Werkov/systemd/tree/freezer-reading-events-expanded) and [freezer-reading-events](https://github.com/Werkov/systemd/tree/freezer-reading-events). The latter is a rebase of the -expanded branch to squash the changes together. I didn't test the behavior.

Wouldn't it be better to use the squashed version (when this PR isn't merged yet), so that it makes backports more readable?

cdown replied:
Let me know once this is resolved and I'll be happy to review again :-)

msekletar commented:
@cdown I've force pushed my branch.

I've addressed (hopefully) everything that you've pointed out, except two points.
- #if HAVE_SELINUX - see the rationale in the commit message
- message that unit type doesn't support freezing - I propose to adjust the message if it proves to be confusing (unless you block merging before we change it :)).

Thank you very much for the review! I owe you a🍺 or two.

msekletar commented:
Ubuntu CI failures seem unrelated (most likely caused by #13794).

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -7687,6 +7747,8 @@ static int systemctl_help(void) {
                "  clean UNIT...                       Clean runtime, cache, state, logs or\n"
                "                                      or configuration of unit\n"
+               "  freeze UNIT...                      Freeze a unit using cgroup freezer\n"
+               "  thaw UNIT...                        Resume execution of a frozen unit\n"
```
In the man page, PATTERN is used, and UNIT here. Please make it consistent.

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -3923,6 +3923,58 @@ static int clean_unit(int argc, char *argv[], void *userdata) {
+
+                r = sd_bus_call(bus, m, 0, &error, NULL);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to freeze unit %s: %s", *name, bus_error_message(&error, r));
```
The error message might be for "thaw" too.

keszybz commented on src/core/slice.c:
```diff
@@ -347,6 +348,55 @@ static void slice_enumerate_perpetual(Manager *m) {
+                if (action == FREEZER_FREEZE && UNIT_VTABLE(u)->freeze)
+                        UNIT_VTABLE(u)->freeze(u);
+                else if (action == FREEZER_THAW && UNIT_VTABLE(u)->thaw)
+                        UNIT_VTABLE(u)->thaw(u);
```
This does nothing for the children that do not support the operation. I'm not sure if that is OK. We would refuse freeze/thaw operations for unsupported unit types, but allow them on the parent slice, and do nothing. Wouldn't it be more appropriate to first loop over all children, check if they support the method, and refuse the whole operation if it cannot be done?

msekletar replied:
OK, I will recursively check the slice before starting the operation on child units.

poettering commented:
(btw, would be great to mark the issues raised here as "resolved" when they are addressed in a new version of the PR, to unclutter the view here).

@msekletar ping, could you rebase this?

msekletar commented:
I am working on this ATM. Not quite sure whether I shouldn't open a new PR once I have the new revision ready 🤔

poettering commented:
i think it would be ok to keep the same PR, and just click "Resolve conversation" on all points raised here...

msekletar commented on src/core/cgroup.c:
```diff
@@ -2928,6 +2928,55 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+                        if (sscanf(p, "%d", &state) != 1)
+                                return -EIO;
+
+                        return state == 1 ? CGROUP_FROZEN : CGROUP_THAWED;
```
I will use the approach suggested by @Werkov which should make us resistant to future additions to the file

msekletar commented on man/systemctl.xml:
```diff
@@ -303,6 +303,29 @@ Sun 2017-02-26 20:57:49 EST  2h 3min left  Sun 2017-02-26 11:56:36 EST  6h ago
+            command line.</para>
+
+            <para>This is the inverse operation to the <command>freeze</command> command and resumes the execution of
+            processes in the unit's cgroup.</para>
```
Yes. At least it behaved like that in my testing.

msekletar commented:
@poettering I think I've addressed all the issue that has been pointed out so far so I've pushed a new version. PTAL.

poettering commented on src/core/cgroup.c:
```diff
@@ -2769,6 +2781,36 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+
+        r = cg_get_keyed_attribute(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, "cgroup.events", STRV_MAKE("populated", "frozen"), values);
+        /* "frozen" key exists only in kernels >v5.2 so assume thawed cgroup before that. */
+        if (r != ENXIO && r < 0)
```
ENXIO → -ENXIO

poettering replied:
hmm, if the call fails because an attribute is missing then ENXIO is returned and the function fails and doesn't touch `values`. This means on old kernels that have the populated field, but do not have the frozen field this will fail and not the populated state will not be iitialized either, and that's a problem...

not sure what the best approach here is... maybe just change cg_get_keyed_attribute() so that instead of returning ENXIO if any specified attribute is missing it initializes missing attributes to NULL + returns the number of initialized fields as return value. Our current users of cg_get_keyed_attributes() all pass a single attribute only so far, hence it should be straight-forward to change the call like that. Of course that means all users like that should check the return value carefully then and consider a return value of 0 like they prevously considered -ENXIO, and only > 0 means "field properly found".

poettering commented on src/core/cgroup.c:
```diff
@@ -2769,6 +2781,36 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+
+        assert(u);
+
+        r = cg_get_keyed_attribute(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, "cgroup.events", STRV_MAKE("populated", "frozen"), values);
```
could you line break this a bit?

poettering commented on src/core/cgroup.c:
```diff
@@ -2769,6 +2781,36 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+        /* The cgroup.events notifications can be merged together so act as we saw the given state for the
+         * first time. The functions we call to handle given state are idempotent, which makes them
+         * effectively remember the previous state. */
+        if (values[0] == NULL || streq(values[0], "1"))
```
doesn't matter, but we usually check strings for != NULL by relying on C's downgrade-to-bool feature, i.e. maybe:

```c
if (!values[0] || streq(values[0], "1")
``
Hmm, if the "populated" field is missing, maybe we should not change the cgroup empty queue at all, as we have no knowledge, and the empty queue processing verifies empty state anyway when it is dispatched... i.e. hence maybe:

```c
if (values[0]) {
        if (streq(values[0], "1")
                 …
         else
              …
}
```

poettering commented on src/core/cgroup.c:
```diff
@@ -2769,6 +2781,36 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+        else
+                unit_add_to_cgroup_empty_queue(u);
+
+        if (values[1] == NULL || streq(values[1], "0"))
```
drop == NULL

poettering commented on src/core/cgroup.h:
```diff
@@ -47,6 +47,23 @@ typedef enum CGroupDevicePolicy {
+
+        _CGROUP_CONTROL_EVENT_MAX,
+        _CGROUP_CONTROL_EVENT_INVALID = -1
+} CGroupControlEvent;
```
this is not used anymore, is it?

poettering commented on src/core/dbus-unit.c:
```diff
@@ -722,6 +723,53 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_new_method_return(message, &u->pending_freezer_message);
```
hmm, we blindly override `pending_freezer_message` here. We should refuse operation if the field is already initialized, i.e. if there's already a freeze operation pending.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4262,13 +4315,18 @@ static void print_status_info(
+        if (ss) {
+                if (!streq(i->freezer_state, "running"))
+                        printf("   Active: %s%s (%s)%s",
+                               active_on, strna(i->active_state), i->freezer_state, active_off);
```
this hides the substate if we have a freezer state. i think we should show all three states in this case

poettering commented on src/core/unit.c:
```diff
@@ -1898,6 +1906,8 @@ int unit_reload(Unit *u) {
                 return 0;
         }
 
+        unit_cgroup_freezer_action(u, FREEZER_THAW);
```
ditto?

poettering commented on src/core/slice.c:
```diff
@@ -347,6 +348,95 @@ static void slice_enumerate_perpetual(Manager *m) {
+                        return 0;
+        }
+
+        return 1;
```
hmm, if we want to determine the units inside a slice we generally enumerate the units with a UNIT_AFTER dependency and then check if their slice field matches our slice. This works because assigning a unit to a slice implies such a dependency. This is a lot better to check than the actual cgroups structure, because the cgroup structure is effect of our structure but not the structure itself, if you follow what I mean

Werkov replied:
I agree that traversing the internal structures is better than traversing the artifact (e.g. current approach would break on delegated units).
That got me to a question -- is the recursion necessary at all? The kernel behavior is already hierarchical, i.e. it would suffice to freeze the chosen cgroup and update respective unit's descendants based on event notifications.

poettering replied:
We should track the freeze state in each unit, hence we probably should propagate this down the tree in our code. But you are right, we might not need to trigger this in the kernel on each level, but it should be sufficient to trigger that only in the top slice we operate on and let the kernel propagate this down.

poettering commented on src/core/slice.c:
```diff
@@ -347,6 +348,95 @@ static void slice_enumerate_perpetual(Manager *m) {
+        else if (r == 0)
+                return log_unit_warning_errno(s, -EOPNOTSUPP, "Requested freezer operation is not supported by all children of the slice");
+
+        r = cg_enumerate_subgroups(SYSTEMD_CGROUP_CONTROLLER, s->cgroup_path, &d);
```
as above, let's enumerate via dependency relationships, not cgroups

Werkov commented on src/core/slice.c:
```diff
@@ -347,6 +348,95 @@ static void slice_enumerate_perpetual(Manager *m) {
+                        r = slice_freezer_action_supported_by_children(u);
+                        if (r <= 0)
+                                return r;
+                } else if (!UNIT_VTABLE(u)->freeze || !UNIT_VTABLE(u)->thaw)
```
I see that `unit_freezer_action` does some logical checks on freezing suitability (e.g. no running job), freezing via slice seems to bypass this conditions.

poettering commented:
So I think we need to figure out what recursive freezing means. In particular what happens if we manage some subtree only, and whoever manages things further up the tree then decides to freeze our whole thing.

It is my understanding that the "frozen" property always reflects the actual frozen state, i.e. it reports the effect of both the freezing of the unit itself and of the freezing of any of the parent cgroups. Is that correct?

If it is I think we should decouple the FreezeState variable we maintain for a unit from what we find reported in the cgroup to some degree. Specifically, our own freeze state should only reflect what the user asked us directly to do with the unit itself and not what might have happened to us because of some upper slic eunit or even some cgroup manager further up the tree. this means we should care for the frozen state reported by the cgroup only while transitioning between states, but not otherwise. i.e. if a unit's cgroup is frozen because we run in a container and the container manager froze the whole container then we should not reflect in the state we maintain for the unit — its state should only
reflect what the user asked us to do about the unit itself.

That said, we should probably have a call unit_freeze_state_recursive() that is the same as unit_freeze_state() but goes recursively up the tree reporting "freezing"/"frozen" if any parent slice says so. Then we should report that via dbus, but purely for display reasons, not for our own control, if you follow what I mean?

That also means, we never actually propagate the freeze state ourselves *down* the tree. Instead the freeze state is always managed for a specific unit only, but possibly reported recursively *up* the tree, if you follow what I mean?

msekletar commented:
> So I think we need to figure out what recursive freezing means. In particular what happens if we manage some subtree only, and whoever manages things further up the tree then decides to freeze our whole thing.

With current implementation we would report "frozen" state for some units and "running" for others, depending on when exactly we get stopped (i.e. notifications we get to process before we get frozen). But, this is mostly moot since we will be stopped and can't report anything to anyone.

> It is my understanding that the "frozen" property always reflects the actual frozen state, i.e. it reports the effect of both the freezing of the unit itself and of the freezing of any of the parent cgroups. Is that correct?

Yes, it reports actual state as well as "transitional" states. Also, there can be unsolicited updates of the actual state, e.g. when service slice is frozen then service's state will be updated too.

> That also means, we never actually propagate the freeze state ourselves down the tree. Instead the freeze state is always managed for a specific unit only, but possibly reported recursively up the tree, if you follow what I mean?

I understand what you mean, however, I fail to see the benefit of doing it this way. Is there anything wrong with how this stuff works currently? I am happy to rework things as needed but I'd like to better understand why I am doing the rework in the first place.

msekletar commented on src/core/unit.c:
```diff
@@ -1790,6 +1796,7 @@ int unit_start(Unit *u) {
          * waits for a holdoff timer to elapse before it will start again. */
 
         unit_add_to_dbus_queue(u);
+        unit_cgroup_freezer_action(u, FREEZER_THAW);
```
We discussed this offline and came to the conclusion that kernel won't return EBUSY.

msekletar commented:
I've force pushed my branch. I think I've addressed all the issues that you've pointed out so far. @keszybz @poettering @Werkov PTAL

@poettering @Werkov regarding recursive behaviour, it might make sense to have a look at integration test that I've wrote and let me know if the current semantics are OK

mrc0mmand commented on src/core/slice.c:
```diff
@@ -347,6 +348,91 @@ static void slice_enumerate_perpetual(Manager *m) {
+}
+
+static int slice_freezer_action(Unit *s, FreezerAction action) {
+        _cleanup_closedir_ DIR *d = NULL;
```
```
    archlinux_systemd_ci: ../src/core/slice.c:379:33: error: unused variable 'd' [-Werror,-Wunused-variable]
    archlinux_systemd_ci:         _cleanup_closedir_ DIR *d = NULL;
    archlinux_systemd_ci:                                 ^
```

mrc0mmand commented on src/core/slice.c:
```diff
@@ -347,6 +348,91 @@ static void slice_enumerate_perpetual(Manager *m) {
 }
 
+static int slice_freezer_action_supported_by_children(Unit *s) {
+        _cleanup_closedir_ DIR *d = NULL;
```
```
    archlinux_systemd_ci: ../src/core/slice.c:352:33: error: unused variable 'd' [-Werror,-Wunused-variable]
    archlinux_systemd_ci:         _cleanup_closedir_ DIR *d = NULL;
    archlinux_systemd_ci:                                 ^
    archlinux_systemd_ci: 2 errors generated.
```

msekletar commented:
@mrc0mmand fixed. Thank you!

msekletar commented:
Hey @keszybz, thanks for the review! I've addressed all your comments and pushed the new version.

keszybz commented:
TEST-38-FREEZER failed in centos ci (arch in kvm) and (centos 7), and bionic-amd64 and bionic-ppc64el. I don't see the exact reason, but it seems related since the test is being added in this PR.

bionic-i386 fails with a timeout. This happens often and seems unrelated.

keszybz commented on src/basic/cgroup-util.c:
```diff
@@ -1704,17 +1704,16 @@ int cg_get_keyed_attribute(
 
-done:
-        memcpy(ret_values, v, sizeof(char*) * n);
-        return 0;
```
Thanks, it's crystal clear now ;)

keszybz commented on src/core/cgroup.h:
```diff
@@ -47,6 +47,14 @@ typedef enum CGroupDevicePolicy {
 
+typedef enum FreezerAction {
+        FREEZER_THAW,
+        FREEZER_FREEZE,
```
I know this is close to OCD, but since I think another push will be necessary anyway, please bear with me ;)

"freeze" happens before "thaw", so the order of definition would be better reversed.

keszybz commented on test/TEST-38-FREEZER/testsuite-38.service:
```diff
@@ -0,0 +1,7 @@
+[Service]
+ExecStart=/usr/lib/systemd/tests/testdata/units/%N.sh
+Type=oneshot
+EOF
```
Remove "EOF".

keszybz commented on test/TEST-38-FREEZER/testsuite.sh:
```diff
@@ -0,0 +1,308 @@
+systemd-analyze log-target console
+
+unit=foo.service
+unit_file=/etc/systemd/system/${unit}
```
This name is fixed. Creating the unit dynamically is not exactly wrong, but OTOH, I don't think we gain anything by this complication. Please just rename this to e.g. `testsuite-38-foo.service` and put it in a static file alongside `testsuite-38.service`.

keszybz commented on test/TEST-38-FREEZER/testsuite.sh:
```diff
@@ -0,0 +1,308 @@
+    systemctl stop "${unit}" >/dev/null 2>&1
+    rm -f "${unit_file}"
+    systemctl daemon-reload
+}
```
This cleanup is not needed. TEST-* doesn't have to leave the machine in pristine state (as long as everything goes away after after a reboot).

keszybz commented on test/test-functions:
```diff
@@ -926,6 +926,7 @@ install_basic_tools() {
     dracut_install -o sushell
     # in Debian ldconfig is just a shell script wrapper around ldconfig.real
     dracut_install -o ldconfig.real
+    dracut_install mktemp cut tr timeout
```
There is a list of tools — one per line, alphabetically sorted — elsewhere in this file. Please add there.

msekletar commented:
I think I've resolved all the issues that you've pointed out. @keszybz PTAL

keszybz commented on src/core/dbus-manager.c:
```diff
@@ -2493,6 +2501,8 @@ const sd_bus_vtable bus_manager_vtable[] = {
         SD_BUS_METHOD("KillUnit", "ssi", NULL, method_kill_unit, SD_BUS_VTABLE_UNPRIVILEGED),
         SD_BUS_METHOD("CleanUnit", "sas", NULL, method_clean_unit, SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD("FreezeUnit", "s", NULL, method_freeze_unit, SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD("ThawUnit", "s", NULL, method_thaw_unit, SD_BUS_VTABLE_UNPRIVILEGED),
```
This now needs to be:
```C
       SD_BUS_METHOD_WITH_NAMES("FreezeUnit",
                                 "s",
                                 SD_BUS_PARAM(name),
                                 NULL,,
                                 method_freeze_unit,
                                 SD_BUS_VTABLE_UNPRIVILEGED),
        SD_BUS_METHOD_WITH_NAMES("ThawUnit",
                                 "s",
                                 SD_BUS_PARAM(name),
                                 NULL,,
                                 method_thaw_unit,
                                 SD_BUS_VTABLE_UNPRIVILEGED),
```

keszybz commented on src/core/dbus-unit.c:
```diff
@@ -857,6 +930,8 @@ const sd_bus_vtable bus_unit_vtable[] = {
         SD_BUS_METHOD("Unref", NULL, NULL, bus_unit_method_unref, SD_BUS_VTABLE_UNPRIVILEGED),
         SD_BUS_METHOD("Clean", "as", NULL, bus_unit_method_clean, SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD("Freeze", NULL, NULL, bus_unit_method_freeze, SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD("Thaw", NULL, NULL, bus_unit_method_thaw, SD_BUS_VTABLE_UNPRIVILEGED),
```
... this one is OK because it doesn't take any parameters.

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4275,13 +4286,18 @@ static void print_status_info(
-                       active_on, strna(i->active_state), ss, active_off);
-        else
+        if (ss) {
+                if (!streq(i->freezer_state, "running"))
```
This segfaults when running against an older pid1. `i->freezer_state` must be assumed to be unavailable.

msekletar replied:
Hmm, do we support mix and match of systemctl and systemd versions?

keszybz replied:
Yes. As mentioned on the chat, `-H` is an obvious example, but more generally, we don't require services to be immediately restarted after upgrade, so we must support version mismatches. In particular, we have no mechanism to restart `user@.service` instances, so after a `dnf ugprade`, a version mismatch may easily persist for months.

msekletar replied:
This should be fixed in the latest revision. PTAL.

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4275,13 +4286,18 @@ static void print_status_info(
+                               active_on, strna(i->active_state), ss, i->freezer_state, active_off);
+                else
+                        printf("   Active: %s%s (%s)%s",
+                               active_on, strna(i->active_state), ss, active_off);
```
Maybe make "(frozen)" yellow or orange to make it stand out more?

msekletar replied:
Done

msekletar commented:
@keszybz the result of your test was indeed unexpected. The problem was that I wasn't checking actual state of the freezer subsystem on all code paths. I've moved the check down to `unit_cgroup_freezer_action()` and your example now works as expected (note that after requesting the thaw on `user@1000.service` the service will still remain frozen since parent slice is still frozen).

msekletar commented:
> It would be great if the man page for systemctl specified what happens when freezing and thawing is nested, and when repeated freeze and thaw requests happen.

I'd like to get final ACK on the proposed semantics from @poettering and then I'd be happy to add that.

msekletar commented:
Hmm, autopkgtests on i386 timeout for some reason. Given that timeout occurred only on i386 it seems like some infra flake.

keszybz commented on src/core/dbus-unit.c:
```diff
@@ -724,6 +726,75 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
+        if (r == -EOPNOTSUPP)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Unit '%s' does not support freezing.", u->id);
+        if (r == -EBUSY)
+                return sd_bus_error_setf(error, BUS_ERROR_UNIT_BUSY, "Unit is inactive or has a pending job.");
```
I think it'd be much nicer to split the two error conditions. When testing this, when I see the message, I completely miss the "inactive" part and immediately jump to "what pending job?!!!". Maybe use EHOSTDOWN or similar for "inactive"?

msekletar commented on src/core/cgroup.c:
```diff
@@ -2761,6 +2773,41 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+                        unit_add_to_cgroup_empty_queue(u);
+        }
+
+        /* Disregard freezer state changes due to operations not initiated by us */
```
Current behaviour is a an attempt to to implement,

https://github.com/systemd/systemd/pull/13512#issuecomment-573007207

The think the main idea suggested by Lennart is that managers higher up in the hierarchy always (un)freeze entire subtree. Hence the manager running in delegated subtree doesn't need to care about actions initiated from outside. Since after freeze followed by unfreeze its state will be consistent.  

Also if we would act on external events we might end-up unfreezing units that should remain frozen. For example consider `foo.service` with parent `bar.slice`. If we get thaw on bar.slice due to an external "thaw" on its parent slice we would go ahead and unfreeze everything that is under `bar.slice` including `foo.service`. However, we should really only unfreeze `foo.service` if our client (i.e. on from subtree that is managed by us) ask as to do so.

Previous issue is fixable by tracking two state variables, but since Lennart doesn't want us to care about external events anyway I implemented just that which I think also sidesteps the above issue (covered by explicit test case BTW).

msekletar commented:
Updated. @keszybz @poettering PTAL.

msekletar commented on src/core/slice.c:
```diff
@@ -347,6 +348,89 @@ static void slice_enumerate_perpetual(Manager *m) {
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_THAW));
+
+        r = slice_freezer_action_supported_by_children(s);
+        if (!r)
```
I am an idiot, I’ve changed the function to return bool previously but didn’t fix the call site. Sorry about that...

msekletar commented:
Updated. @poettering PTAL

keszybz commented on src/basic/cgroup-util.c:
```diff
@@ -149,6 +149,20 @@ bool cg_ns_supported(void) {
         return enabled;
 }
 
+bool cg_freezer_supported(void) {
```
Oh, this should go in shared/, not /basic/.

poettering replied:
why? sounds generic enough and similar enough to many other cg_xyz() checkers to stay here...

keszybz commented on src/basic/unit-def.c:
```diff
@@ -108,6 +108,15 @@ static const char* const unit_active_state_table[_UNIT_ACTIVE_STATE_MAX] = {
+        [FREEZER_THAWING] = "thawing",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(freezer_state, FreezerState);
```
This too.

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1749,7 +1750,10 @@ int cg_get_keyed_attribute(
+        if (mode & CG_KEY_MODE_GRACEFUL)
+                goto done;
+        else
+                r = -ENXIO;
```
doesn't really matter, but the "else" can just be removed. That way we save one line and a bit of indentation on this line...

poettering commented on src/core/cgroup.c:
```diff
@@ -2761,6 +2773,41 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
+        assert(u);
+
+        r = cg_get_keyed_attribute(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, "cgroup.events",
+                                   STRV_MAKE("populated", "frozen"), values);
```
didn't you want to use graceful mode here?

msekletar replied:
Oh god...fixed.

poettering commented on src/core/cgroup.c:
```diff
@@ -3550,6 +3597,45 @@ int compare_job_priority(const void *a, const void *b) {
+        else
+                u->freezer_state = FREEZER_THAWING;
+
+
```
spurious double newline

poettering commented on src/core/unit.c:
```diff
@@ -6076,6 +6125,80 @@ int unit_can_clean(Unit *u, ExecCleanMask *ret) {
+                return -EHOSTDOWN;
+
+        s = unit_active_state(u);
+        if (!IN_SET(s, UNIT_ACTIVE))
```
no point in using IN_SET here. a simple `!=` check suffices

poettering commented:
did another review round. found some more minor stuff, ptal

msekletar commented on src/core/dbus-unit.c:
```diff
@@ -719,7 +719,7 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
                 return sd_bus_error_setf(error, BUS_ERROR_NOTHING_TO_CLEAN, "No matching resources found.");
         if (r == -EBUSY)
-                return sd_bus_error_setf(error, BUS_ERROR_UNIT_BUSY, "Unit is not inactive or has pending job.");
+                return sd_bus_error_setf(error, BUS_ERROR_UNIT_BUSY, "Unit is inactive or has pending job.");
```
I've dropped the commit that introduced the change. I got confused by double negative and as I've never played with clean so far it didn't occur to me it is actually correct.

keszybz replied:
The grammar is so tortuous that I was confused too (I suggested the change).

msekletar commented:
Fixed. @poettering PTAL.

poettering commented:
lgtm. Now I hope the CIs can be convinced too!

msekletar commented:
@poettering thanks for the reviews!

Btw, I did one more fix-up to make Travis clang (asan+ubsan) job green (hopefully).

msekletar commented:
Ubuntu CI failure seems unrelated, from the log it looks like some Openstack issue. Travis CI failure also seems unrelated (docker exec failure).

mrc0mmand commented:
> Ubuntu CI failure seems unrelated, from the log it looks like some Openstack issue. Travis CI failure also seems unrelated (docker exec failure).

I restarted at least the Travis CI job.

Werkov commented:
Hi, good job pulling this off. I only got down to review this now, so let me know if you prefer followup here or would this better be filed as issue(s) against `master`.

### Freezing vs jobs

This manifests in cases like: `systemd-run --slice=frozen.slice cmd` or `systemctl stop unit-frozen-by-ancestor.slice` (the latter mainly visible with `ExecStop=`). At one moment I considered making freezing/thawing a first class type of job, however, that wouldn't compose nicely with established state machine (active, inactive, ...). In the end, I think it'd be simpler to reject jobs (basically anything but JOB_DONE to fail deps) on units that are currently in frozen context (regardless of freezing origin, i.e. frozen ancestor or self).

### Recursive behavior

(I recall discussing this but can't find that now and it may be less relevant to the current version anyway.)
The behavior I observe is

> : nothing frozen at the beginning
> systemctl freeze child.service # effectively frozen, reported state is frozen
> systemctl freeze parent.slice   # effectively frozen, reported state is frozen
> systemctl thaw parent.slice       # everything is thawed

or

> : nothing frozen at the beginning
> systemctl freeze parent.slice
> : child.service effectively frozen, reported state is frozen
> systemctl thaw child.service # effectively frozen, reported state remains frozen
> systemctl thaw child.service # effectively frozen, reported state switches to thawing

In the former case, it seems to be that user's request to have child.service frozen was wiped by thawing parent.slice. [1]

In the latter case, thawing the child.service correctly doesn't change reported state from frozen (because of the parent), however, the change after second call is weird. 

I'd expect the following behavior
> : nothing frozen at the beginning
> systemctl freeze child.service # effectively frozen, reported state is frozen
> systemctl freeze parent.slice   # effectively frozen, reported state is frozen
> systemctl thaw parent.slice      # child effectively frozen, reported state is frozen

> : nothing frozen at the beginning
> systemctl freeze parent.slice
> : child.service effectively frozen, reported state is frozen-ancestor
> systemctl thaw child.service # effectively frozen, reported state remains frozen-ancestor
> ...
> systemctl freeze child.service # effectively frozen, reported state switches to frozen


[1] Although lexically substituting freeze/thaw with start/stop would make the behavior familiar (in a sense that child.service can be started via dependency by parent.slice even though user requested child.service to stop), I think it's not what is intended when configuring properties in the hierarchy. I'd like to know your opinion.

### Freezing vs jobs 2

During my experiments (with 9e12d5bf63) I run into  the following situation. I didn't catch the steps to reproduce it, so it's rather a hint.
> image:/etc/systemd/system # systemctl status test.slice
> ● test.slice
>      Loaded: loaded
>      Active: active since Wed 2020-05-06 18:09:32 CEST; 21s ago
>       Tasks: 4
>      Memory: 3.2M
>         CPU: 75ms
>      CGroup: /test.slice
>              ├─looper10.service
>              │ ├─495 /usr/bin/bash /root/looper 10
>              │ └─537 sleep 1
>              └─looper11.service
>                ├─502 /usr/bin/bash /root/looper 11
>                └─536 sleep 1
>
> May 06 18:09:49 image looper[502]: s11
> May 06 18:09:49 image looper[495]: s10
> May 06 18:09:50 image looper[502]: s11
> May 06 18:09:50 image looper[495]: s10
> May 06 18:09:51 image looper[502]: s11
> May 06 18:09:51 image looper[495]: s10
> May 06 18:09:52 image looper[502]: s11
> May 06 18:09:52 image looper[495]: s10
> May 06 18:09:53 image looper[502]: s11
> May 06 18:09:53 image looper[495]: s10
> image:/etc/systemd/system # systemctl list-jobs
> No jobs running.                                                                                                                                                                                
> image:/etc/systemd/system # systemctl freeze test.slice
> Failed to freeze unit test.slice: Unit has a pending job.

### Minor remarks

I suggest making freezing/thawing messages same logging level as starting/stopping. I can imagine troubleshooting some issues with regular logging may get cumbersome.

I ran into misaligned coloring, [this](https://github.com/Werkov/systemd/commit/14574dff8c7691400e044b460e24240fe550d648) should fix it.

msekletar commented:
@Werkov it would be much appreciated if you could file separate issue. Thanks!

Werkov commented:
I've filed #15849, #15850 and #15851, I skipped *Freezing vs jobs 2* for the lack of data.
A bit more time passed than I expected but the points should still be valid.

venuprsd commented:
Hi Michal

Hope you are doing good. I use RHEL 8.3. My systemd still getting error "thawing". Any official fix? Please help me

venuprsd commented:
> Hi Michal
> 
> Hope you are doing good. I use RHEL 8.3. My systemd still getting error "thawing". Any official fix? Please help me

msekletar commented:
> Hope you are doing good. I use RHEL 8.3. My systemd still getting error "thawing". Any official fix? Please help me

Please update to latest RHEL-8.3 systemd package (systemd-239-41.el8_3).
